### PR TITLE
Jobset dynamic reclaim

### LINKED ISSUE #820: [JobSet] Support partial admission
**What would you like to be added**:

Support for partial admission in JobSet https://github.com/kubernetes-sigs/kueue/tree/main/keps/420-partial-admission

The open questions are:
- How to support resizing of multiple PodSets? Maximizing the size of one podset might imply reducing the size of another. Even if we have some priority to maximize the size of specific podsets, the problem is generally NP, so we might need to reduce the search space, or just limit the support to one PodSet.
- What to do if a Job is replicated? In this case, there are multiple options for sizing. But perhaps a replicated Job always has full size.

**Why is this needed**:

To be able to admit jobs of smaller size if this is acceptable to the application.

**Completion requirements**:

This enhancement requires the following artifacts:

- [x] Design doc
- [x] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Adds support for dynamic reclaim in the JobSet integration.
#### Which issue(s) this PR fixes:

Fixes #820 

#### Does this PR introduce a user-facing change?

```release-note
Add support for dynamic reclaim in the JobSet integration.
```

### CODE REVIEW

**Comment by trasc** [comment]
/hold 
for #762

File: pkg/controller/jobs/jobset/jobset_controller.go [hunk 1]
```diff
@@ -182,6 +184,31 @@ func (j *JobSet) PodsReady() bool {
...
+
+	ret := make([]kueue.ReclaimablePod, 0, len(j.Spec.ReplicatedJobs))
+	specs := slices.ToRefMap(j.Spec.ReplicatedJobs, func(js *jobsetapi.ReplicatedJob) string { return js.Name })
+	statuses := slices.ToRefMap(j.Status.ReplicatedJobsStatus, func(js *jobsetapi.ReplicatedJobStatus) string { return js.Name })
+
+	for name, spec := range specs {
+		if status, found := statuses[name]; found && status.Succeeded > 0 {
+			pc := podsCount(spec)
+			totalCount := pointer.Int32Deref(spec.Template.Spec.Completions, pointer.Int32Deref(spec.Template.Spec.Parallelism, 1)) * int32(spec.Replicas)
+			remaining := totalCount - status.Succeeded
```
Comment (alculquicondor) [inline pkg/controller/jobs/jobset/jobset_controller.go:200]: is `status.Succeeded` the number of successful pods across the replicated jobs?

File: pkg/controller/jobs/jobset/jobset_controller.go [hunk 2]
```diff
@@ -182,6 +184,31 @@ func (j *JobSet) PodsReady() bool {
 	return replicas == readyReplicas
 }
 
+func (j *JobSet) ReclaimablePods() []kueue.ReclaimablePod {
```
Comment (alculquicondor) [inline pkg/controller/jobs/jobset/jobset_controller.go:187]: let's add unit tests (with a test table) for this.

File: test/integration/controller/jobset/jobset_controller_test.go [hunk 3]
```diff
@@ -672,4 +672,108 @@ var _ = ginkgo.Describe("JobSet controller interacting with scheduler", ginkgo.O
...
+			testingjobset.ReplicatedJobRequirements{
+				Name:        "replicated-job-1",
+				Replicas:    2,
+				Parallelism: 1,
+				Completions: 1,
+			}, testingjobset.ReplicatedJobRequirements{
+				Name:        "replicated-job-2",
+				Replicas:    3,
+				Parallelism: 1,
+				Completions: 1,
```
Comment (alculquicondor) [inline test/integration/controller/jobset/jobset_controller_test.go:692]: let's have higher parallelism here

File: test/integration/controller/jobset/jobset_controller_test.go [hunk 4]
```diff
@@ -672,4 +672,108 @@ var _ = ginkgo.Describe("JobSet controller interacting with scheduler", ginkgo.O
...
+			createdJobSet2 := &jobsetapi.JobSet{}
+			gomega.Consistently(func() *bool {
+				gomega.Expect(k8sClient.Get(ctx, lookupKey2, createdJobSet2)).Should(gomega.Succeed())
+				return createdJobSet2.Spec.Suspend
+			}, util.ConsistentDuration, util.Interval).Should(gomega.Equal(pointer.Bool(true)))
+			util.ExpectPendingWorkloadsMetric(clusterQueue, 0, 1)
+			util.ExpectAdmittedActiveWorkloadsMetric(clusterQueue, 1)
+		})
+
+		ginkgo.By("checking the second job starts when the first has less then to completions to go", func() {
```
Comment (alculquicondor) [inline test/integration/controller/jobset/jobset_controller_test.go:741]:
```suggestion
		ginkgo.By("checking the second job starts when the first has less than two completions to go", func() {
```
?

File: pkg/controller/jobs/jobset/jobset_controller.go (hunk 2 above)
Comment (trasc) [reply pkg/controller/jobs/jobset/jobset_controller.go:187]: done

File: pkg/controller/jobs/jobset/jobset_controller.go (hunk 1 above)
Comment (trasc) [reply pkg/controller/jobs/jobset/jobset_controller.go]: apparently no, reworked

File: test/integration/controller/jobset/jobset_controller_test.go (hunk 3 above)
Comment (trasc) [reply test/integration/controller/jobset/jobset_controller_test.go]: done

File: test/integration/controller/jobset/jobset_controller_test.go (hunk 4 above)
Comment (trasc) [reply test/integration/controller/jobset/jobset_controller_test.go]: done

File: pkg/controller/jobs/jobset/jobset_controller.go [hunk 5]
```diff
@@ -182,16 +184,41 @@ func (j *JobSet) PodsReady() bool {
...
+	if comp := pointer.Int32Deref(spec.Completions, jobPodsCount); comp < jobPodsCount {
 		jobPodsCount = comp
 	}
+	return jobPodsCount
+}
+
+func podsCount(rj *jobsetapi.ReplicatedJob) int32 {
 	// The JobSet's operator validates that this will not overflow.
-	return int32(replicas) * jobPodsCount
+	return int32(rj.Replicas) * podsCountPerReplica(rj)
```
Comment (alculquicondor) [inline pkg/controller/jobs/jobset/jobset_controller.go:221]: Opened https://github.com/kubernetes-sigs/jobset/issues/218

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**